### Ashlyn Strickland
#### Group 6
#### Project: Reproducibility
#### 3/31/2025

In [1]:
## Section 1 

import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score, classification_report

from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB

### import needed libraries

In [2]:
## Section 2

xuetangx_df = pd.read_csv("~/Desktop/DATA 6550/all_normalized_features.csv")

xuetangx_df.head()

### display dataset 

,enroll_id,action_count,seek_video_count,play_video_count,pause_video_count,stop_video_count,load_video_count,problem_get_count,problem_check_count,problem_save_count,...,click_forum_count,click_progress_count,close_courseware_count,truth,age,gender,education,user_enroll_count,course_enroll_count,course_category
0,772,-0.228387,-0.243043,-0.296739,-0.333535,-0.059298,-0.480968,-0.042412,-0.305844,-0.202161,...,-0.17735,-0.284796,-0.464309,1,29,0,3,-0.007581,-0.138544,0
1,773,-0.204883,-0.115253,-0.242027,-0.296012,-0.059298,-0.445267,-0.042412,-0.305844,-0.202161,...,-0.17735,-0.284796,-0.428065,1,0,0,0,-0.360012,-0.138544,0
2,774,-0.139320,-0.051358,-0.105248,-0.208460,-0.057709,-0.195358,-0.042412,-0.305844,-0.202161,...,-0.17735,-0.284796,-0.138111,1,0,0,0,-0.095689,-0.138544,0
3,776,-0.207357,-0.243043,-0.255705,-0.296012,-0.057709,-0.409566,-0.042412,-0.305844,-0.202161,...,-0.17735,-0.284796,-0.391821,0,37,0,6,-0.007581,-0.138544,0
4,777,-0.225912,-0.243043,-0.296739,-0.333535,-0.059298,-0.480968,-0.042412,-0.305844,-0.202161,...,-0.17735,-0.284796,-0.428065,1,0,0,0,0.432957,-0.138544,0


In [3]:
## Section 3

xuetangx_X = xuetangx_df.drop(columns=['truth', 'enroll_id'])
xuetangx_y = xuetangx_df['truth']

### features and target (truth) were separated from the XuetangX dataset

In [4]:
## Section 4

xuetangx_X_train, xuetangx_X_test, xuetangx_y_train, xuetangx_y_test = train_test_split(
    xuetangx_X, xuetangx_y, test_size=0.2, random_state=100)

### the XuetangX data was split into training and testing sets with an 80/20 ratio

In [5]:
## Section 5

x_sample = xuetangx_X_train.sample(n=100, random_state=100)
y_sample = xuetangx_y_train.loc[x_sample.index]

### sample of 100 training records was selected for lightweight model training and testing

In [6]:
## Section 6

rf_model = RandomForestClassifier(n_estimators=10, random_state=100)

rf_model.fit(x_sample, y_sample)
rf_pred = rf_model.predict(xuetangx_X_test)

print("RF Accuracy:", accuracy_score(xuetangx_y_test, rf_pred))
print("\n", classification_report(xuetangx_y_test, rf_pred))

rf_cv = cross_val_score(rf_model, x_sample, y_sample, cv=5, n_jobs=1)
print("RF CV Mean:", np.mean(rf_cv))
print("RF CV Std:", np.std(rf_cv))

### Random Forest model trained on the 100-sample subset and evaluated using test accuracy and 5-fold cross-validation

RF Accuracy: 0.7733386514214806

               precision    recall  f1-score   support

           0       0.52      0.59      0.56     10813
           1       0.87      0.83      0.85     34316

    accuracy                           0.77     45129
   macro avg       0.70      0.71      0.70     45129
weighted avg       0.78      0.77      0.78     45129

RF CV Mean: 0.78
RF CV Std: 0.039999999999999994


### The Random Forest model achieved 77.3% accuracy on the test set, with strong performance on class 1 (F1-score: 0.85). Cross-validation showed a mean accuracy of 78% with a standard deviation of ~4%, indicating consistent performance.

In [7]:
## Section 7

ab_model = AdaBoostClassifier(n_estimators=10, random_state=100)

ab_model.fit(x_sample, y_sample)
ab_pred = ab_model.predict(xuetangx_X_test)

print("AdaBoost Accuracy:", accuracy_score(xuetangx_y_test, ab_pred))
print("\n", classification_report(xuetangx_y_test, ab_pred))

ab_cv = cross_val_score(ab_model, x_sample, y_sample, cv=5, n_jobs=1)
print("AdaBoost CV Mean:", np.mean(ab_cv))
print("AdaBoost CV Std:", np.std(ab_cv))

### An AdaBoost model was trained on the 100-sample subset and evaluated using test accuracy and 5-fold cross-validation

AdaBoost Accuracy: 0.7626359990250172

               precision    recall  f1-score   support

           0       0.51      0.33      0.40     10813
           1       0.81      0.90      0.85     34316

    accuracy                           0.76     45129
   macro avg       0.66      0.61      0.62     45129
weighted avg       0.74      0.76      0.74     45129

AdaBoost CV Mean: 0.7799999999999999
AdaBoost CV Std: 0.024494897427831806


### The AdaBoost model achieved 76.3% test accuracy, performing well on class 1 (F1-score: 0.85) but weaker on class 0. Cross-validation showed a mean accuracy of 78% with a low standard deviation of ~2.4%, indicating stable performance.

In [8]:
## Section 8

dt_model = DecisionTreeClassifier(max_depth=3, random_state=100)

dt_model.fit(x_sample, y_sample)
dt_pred = dt_model.predict(xuetangx_X_test)

print("Decision Tree Accuracy:", accuracy_score(xuetangx_y_test, dt_pred))
print("\n", classification_report(xuetangx_y_test, dt_pred))

dt_cv = cross_val_score(dt_model, x_sample, y_sample, cv=5, n_jobs=1)
print("Decision Tree CV Mean:", np.mean(dt_cv))
print("Decision Tree CV Std:", np.std(dt_cv))

### A Decision Tree model with max depth 3 was trained on the 100-sample subset and evaluated using test accuracy and 5-fold cross-validation.

Decision Tree Accuracy: 0.7653615192005141

               precision    recall  f1-score   support

           0       0.52      0.33      0.40     10813
           1       0.81      0.90      0.85     34316

    accuracy                           0.77     45129
   macro avg       0.66      0.62      0.63     45129
weighted avg       0.74      0.77      0.75     45129

Decision Tree CV Mean: 0.77
Decision Tree CV Std: 0.09797958971132714


### The Decision Tree model achieved 76.5% test accuracy, with strong performance on class 1 (F1-score: 0.85) but limited recall for class 0. Cross-validation gave a 77% mean accuracy with a higher standard deviation of ~9.8%, indicating more variability across folds.

In [9]:
## Section 9

knn_model = KNeighborsClassifier(n_neighbors=3)

knn_model.fit(x_sample, y_sample)
knn_pred = knn_model.predict(xuetangx_X_test)

print("kNN Accuracy:", accuracy_score(xuetangx_y_test, knn_pred))
print("\n", classification_report(xuetangx_y_test, knn_pred))

knn_cv = cross_val_score(knn_model, x_sample, y_sample, cv=5, n_jobs=1)
print("kNN CV Mean:", np.mean(knn_cv))
print("kNN CV Std:", np.std(knn_cv))

### A k-Nearest Neighbors (kNN) model with 3 neighbors was trained on the 100-sample subset and evaluated using test accuracy and 5-fold cross-validation.

/Users/ashlynbrown/anaconda3/lib/python3.10/site-packages/threadpoolctl.py:1010: RuntimeWarning: 
Found Intel OpenMP ('libiomp') and LLVM OpenMP ('libomp') loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)


kNN Accuracy: 0.7722528750914046

               precision    recall  f1-score   support

           0       0.57      0.20      0.30     10813
           1       0.79      0.95      0.86     34316

    accuracy                           0.77     45129
   macro avg       0.68      0.58      0.58     45129
weighted avg       0.74      0.77      0.73     45129

kNN CV Mean: 0.8299999999999998
kNN CV Std: 0.024494897427831747


### The kNN model achieved 77.2% test accuracy, performing very well on class 1 (F1-score: 0.86) but poorly on class 0 (F1-score: 0.30). Cross-validation showed a strong mean accuracy of 83% with a low standard deviation of ~2.4%, indicating consistent performance.

In [10]:
## Section 10

lr_model = LogisticRegression(max_iter=1000, random_state=100)

lr_model.fit(x_sample, y_sample)
lr_pred = lr_model.predict(xuetangx_X_test)

print("LogReg Accuracy:", accuracy_score(xuetangx_y_test, lr_pred))
print("\n", classification_report(xuetangx_y_test, lr_pred))

lr_cv = cross_val_score(lr_model, x_sample, y_sample, cv=5, n_jobs=1)
print("LogReg CV Mean:", np.mean(lr_cv))
print("LogReg CV Std:", np.std(lr_cv))

### A Logistic Regression model was trained on the 100-sample subset with 1000 max iterations and evaluated using test accuracy and 5-fold cross-validation.

LogReg Accuracy: 0.7930377362671452

               precision    recall  f1-score   support

           0       0.69      0.24      0.36     10813
           1       0.80      0.97      0.88     34316

    accuracy                           0.79     45129
   macro avg       0.75      0.61      0.62     45129
weighted avg       0.78      0.79      0.75     45129

LogReg CV Mean: 0.7699999999999999
LogReg CV Std: 0.040000000000000036


### The Logistic Regression model achieved 79.3% test accuracy, with excellent recall for class 1 (0.97) but low recall for class 0 (0.24). Cross-validation yielded a mean accuracy of 77% and a standard deviation of ~4%, showing stable performance.

In [11]:
## Section 11

lda_model = LinearDiscriminantAnalysis()

lda_model.fit(x_sample, y_sample)
lda_pred = lda_model.predict(xuetangx_X_test)

print("LDA Accuracy:", accuracy_score(xuetangx_y_test, lda_pred))
print("\n", classification_report(xuetangx_y_test, lda_pred))

lda_cv = cross_val_score(lda_model, x_sample, y_sample, cv=5, n_jobs=1)
print("LDA CV Mean:", np.mean(lda_cv))
print("LDA CV Std:", np.std(lda_cv))

### A Linear Discriminant Analysis (LDA) model was trained on the 100-sample subset and evaluated using test accuracy and 5-fold cross-validation.

LDA Accuracy: 0.7754437279797912

               precision    recall  f1-score   support

           0       0.55      0.32      0.41     10813
           1       0.81      0.92      0.86     34316

    accuracy                           0.78     45129
   macro avg       0.68      0.62      0.64     45129
weighted avg       0.75      0.78      0.75     45129

LDA CV Mean: 0.8099999999999999
LDA CV Std: 0.05830951894845301


### The LDA model achieved 77.5% test accuracy, performing strongly on class 1 (F1-score: 0.86) but showing limited performance on class 0. Cross-validation resulted in a mean accuracy of 81% with a standard deviation of ~5.8%, indicating moderate variability across folds.

In [12]:
## Section 12

nb_model = GaussianNB()

nb_model.fit(x_sample, y_sample)
nb_pred = nb_model.predict(xuetangx_X_test)

print("Naive Bayes Accuracy:", accuracy_score(xuetangx_y_test, nb_pred))
print("\n", classification_report(xuetangx_y_test, nb_pred))

nb_cv = cross_val_score(nb_model, x_sample, y_sample, cv=5, n_jobs=1)
print("NB CV Mean:", np.mean(nb_cv))
print("NB CV Std:", np.std(nb_cv))

### A Gaussian Naive Bayes model was trained on the 100-sample subset and evaluated using test accuracy and 5-fold cross-validation.

Naive Bayes Accuracy: 0.7842407321234682

               precision    recall  f1-score   support

           0       0.56      0.46      0.51     10813
           1       0.84      0.89      0.86     34316

    accuracy                           0.78     45129
   macro avg       0.70      0.67      0.68     45129
weighted avg       0.77      0.78      0.78     45129

NB CV Mean: 0.7699999999999999
NB CV Std: 0.08124038404635961


### The Naive Bayes model achieved 78.4% test accuracy, with solid performance on class 1 (F1-score: 0.86) and improved performance on class 0 compared to other models (F1-score: 0.51). Cross-validation produced a mean accuracy of 77% with a standard deviation of ~8.1%, indicating slightly higher variability across folds.

In [13]:
## Section 13

results = {
    'Model': [
        'Random Forest', 'AdaBoost', 'Decision Tree', 'kNN',
        'Logistic Regression', 'LDA', 'Naive Bayes'
    ],
    'Test Accuracy': [
        accuracy_score(xuetangx_y_test, rf_pred),
        accuracy_score(xuetangx_y_test, ab_pred),
        accuracy_score(xuetangx_y_test, dt_pred),
        accuracy_score(xuetangx_y_test, knn_pred),
        accuracy_score(xuetangx_y_test, lr_pred),
        accuracy_score(xuetangx_y_test, lda_pred),
        accuracy_score(xuetangx_y_test, nb_pred)
    ],
    'CV Mean Accuracy': [
        np.mean(rf_cv), np.mean(ab_cv), np.mean(dt_cv),
        np.mean(knn_cv), np.mean(lr_cv), np.mean(lda_cv),
        np.mean(nb_cv)
    ],
    'CV Std Dev': [
        np.std(rf_cv), np.std(ab_cv), np.std(dt_cv),
        np.std(knn_cv), np.std(lr_cv), np.std(lda_cv),
        np.std(nb_cv)
    ]
}

results_df = pd.DataFrame(results)
results_df

### This code creates a DataFrame that compares the performance of seven ML models on the XuetangX dataset by compiling their test accuracy, cross-validation mean accuracy, and cross-validation standard deviation into a results table.

,Model,Test Accuracy,CV Mean Accuracy,CV Std Dev
0,Random Forest,0.773339,0.78,0.040000
1,AdaBoost,0.762636,0.78,0.024495
2,Decision Tree,0.765362,0.77,0.097980
3,kNN,0.772253,0.83,0.024495
4,Logistic Regression,0.793038,0.77,0.040000
5,LDA,0.775444,0.81,0.058310
6,Naive Bayes,0.784241,0.77,0.081240


### The results show that all seven models achieved similar test accuracies around 76–79%, with Logistic Regression having the highest test accuracy (79.3%) and kNN achieving the highest cross-validation mean accuracy (83%), while Decision Tree and Naive Bayes showed more variability across folds.

### Overall Summary 

This project focused on reproducing machine learning results from a research paper that predicted student dropout using the XuetangX dataset. I used a preprocessed file called all_normalized_features.csv, where I separated the features from the target column, truth.

The data was split into training and testing sets (80/20). Because of system limitations, I sampled only 100 rows from the training set to avoid memory issues and crashing while still following the paper’s process.

I trained seven machine learning models: Random Forest, AdaBoost, Decision Tree, kNN, Logistic Regression, LDA, and Naive Bayes. Each model was evaluated using test accuracy, classification metrics, and 5-fold cross-validation.

Logistic Regression had the highest test accuracy (79.3%), while kNN had the best CV accuracy (83%). Most models performed well on class 1 (dropouts) but not as well on class 0 (non-dropouts), likely due to class imbalance.

Even though I trained on a smaller sample, the results closely followed the trends in the original paper. The small differences highlight how tough it is to fully reproduce results — especially when working with limited resources.